homeassistant/components/crownstone/entry_manager.py

```diff
@@ -0,0 +1,226 @@
+from homeassistant.const import (
+    CONF_EMAIL,
+    CONF_PASSWORD,
+    CONF_UNIQUE_ID,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import aiohttp_client
+
+from .const import (
+    ATTR_UART,
+    CONF_USB_PATH,
+    CONF_USE_CROWNSTONE_USB,
+    CROWNSTONE_USB,
+    DOMAIN,
+    PLATFORMS,
+    SSE_LISTENERS,
+    UART_LISTENERS,
+)
+from .helpers import get_port
+from .listeners import setup_sse_listeners, setup_uart_listeners
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class CrownstoneEntryManager:
+    """Manage a Crownstone config entry."""
+
+    uart: CrownstoneUart
+    cloud: CrownstoneCloud
+    sse: CrownstoneSSEAsync
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Initialize the hub."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self.listeners: dict[str, Any] = {}
+        self.usb_sphere_id: str | None = None
+
+    async def async_setup(self) -> bool:
+        """
+        Set up a Crownstone config entry.
+
+        This class is a combination of Crownstone cloud, Crownstone SSE and Crownstone uart.
+        Returns True if the setup was successful.
+        """
+        # Setup email and password gained from config flow
+        customer_email = self.config_entry.data[CONF_EMAIL]
+        customer_password = self.config_entry.data[CONF_PASSWORD]
+
+        # Add entry update listener
+        self.config_entry.async_on_unload(
+            self.config_entry.add_update_listener(async_update_entry_options)
+        )
+
+        # Create cloud instance
+        self.cloud = CrownstoneCloud(
+            email=customer_email,
+            password=customer_password,
+            clientsession=aiohttp_client.async_get_clientsession(self.hass),
+        )
+        # Login
+        try:
+            await self.cloud.async_initialize()
+        except CrownstoneAuthenticationError as auth_err:
+            _LOGGER.error(
+                "Auth error during login with type: %s and message: %s",
+                auth_err.type,
+                auth_err.message,
+            )
+            return False
+        except CrownstoneUnknownError as unknown_err:
+            _LOGGER.error("Unknown error during login")
+            raise ConfigEntryNotReady from unknown_err
+
+        # Create SSE instance
+        # a new clientsession is created because the default one does not cleanup on unload
+        self.sse = CrownstoneSSEAsync(
+            email=customer_email,
+            password=customer_password,
+            access_token=self.cloud.access_token,
+            websession=aiohttp_client.async_create_clientsession(self.hass),
+        )
+        # Listen for events in the background, without task tracking
+        asyncio.create_task(self.async_process_events(self.sse))
```

Comment:
Does this need to be canceled when the config entry is unloaded?

Reply:
this is done in `self.sse.close_client()` in `async_unload`. It closes the connection, and exits the async context manager, which finishes the task